lib/parser.ml

```diff
@@ -0,0 +1,324 @@
+      ; num_captures = 1
+      ; extract_value_from_captured_groups = get_first_char_of_group
+      }
+    ;;
+
+    let lower : 'a t =
+      { regex = Re.group (Re.set Re2_compatiblity.lower)
+      ; num_captures = 1
+      ; extract_value_from_captured_groups = get_first_char_of_group
+      }
+    ;;
+
+    let alpha : char t =
+      { regex = Re.group (Re.set Re2_compatiblity.alpha)
+      ; num_captures = 1
+      ; extract_value_from_captured_groups = get_first_char_of_group
+      }
+    ;;
+
+    let alnum : char t =
+      { regex = Re.group (Re.set Re2_compatiblity.alnum)
+      ; num_captures = 1
+      ; extract_value_from_captured_groups = get_first_char_of_group
+      }
+    ;;
+
+    let one_of : char list -> char t =
+     fun chars ->
+      { regex = Re.group (Re.set (String.of_char_list chars))
+      ; num_captures = 1
+      ; extract_value_from_captured_groups = get_first_char_of_group
+      }
+   ;;
+
+    let not_one_of : char list -> char t =
+     fun chars ->
+      { regex = Re.group (Re.compl [ Re.set (String.of_char_list chars) ])
+      ; num_captures = 1
+      ; extract_value_from_captured_groups = get_first_char_of_group
+      }
+   ;;
+
+    (* let digit : 'a t = "\\d"
+    let space : 'a t = "\\s"
+    let upper :  'a t = "[A-Z]"
+    let lower : 'a t = "[a-z]"
+    let alpha : 'a t = "[a-zA-Z]"
+    let alnum 
+     *)
+  end
+
+  let string (s : string) : unit t =
+    { regex = Re.str s
+    ; num_captures = 0
+    ; extract_value_from_captured_groups = (fun _ _ -> Some ())
+    }
+  ;;
+
+  let any_string : string t =
+    { regex = Re.group (Re.rep Re.any)
+    ; num_captures = 1
+    ; extract_value_from_captured_groups = get_first_string_of_group
+    }
+  ;;
+
+  let start_of_input : unit t =
+    { regex = Re.bos
+    ; num_captures = 0
+    ; extract_value_from_captured_groups = (fun _ _ -> Some ())
+    }
+  ;;
+
+  let end_of_input : unit t =
+    { regex = Re.eos
+    ; num_captures = 0
+    ; extract_value_from_captured_groups = (fun _ _ -> Some ())
+    }
+  ;;
+
+  let ignore_m : _ t -> unit t =
+   fun t ->
+    { regex = t.regex
+    ; num_captures = t.num_captures
+    ; extract_value_from_captured_groups = (fun _ _ -> Some ())
+    }
+ ;;
+
+  (* let capture : unit t -> string t =
+   fun t ->
+    { regex = Re.group t.regex
+    ; num_captures = t.num_captures + 1
+    ; extract_value_from_captured_groups = get_first_string_of_group
+    }
+ ;; *)
+
+  let and_capture (t : 'a t) : ('a * string) t =
+    { regex = Re.group t.regex
+    ; num_captures = t.num_captures + 1
+    ; extract_value_from_captured_groups =
+        (fun group offset ->
+          let the_thing_we_captured_ourselves : string option =
+            get_first_string_of_group group offset
+          in
+          let prev_parser_thing : 'a option =
+            t.extract_value_from_captured_groups group (offset + 1)
+          in
+          Option.both prev_parser_thing the_thing_we_captured_ourselves)
```

Comment:
better naming pls

Reply:
renamed as current and previous capture respectively. Is that fine?